### PR TITLE
SkinnedMesh: support weight normalization for BufferGeometry

### LINKED ISSUE #7111: Fog and Whites r72
It seems that whiter colors can pierce through the fog in 72. This was not an issue in 71.

http://jsfiddle.net/titansoftime/nu1217ch/

Changing to the master build provides expected results. This bug makes scenes with fog (most of mine) look kinda odd.

### PR DESCRIPTION
**Pull request by WestLangley**
Fixes #7111 and #7197.

/ping @titansoftime Please have a look.

### CODE REVIEW

File: src/objects/SkinnedMesh.js [hunk 1]
```diff
@@ -109,15 +109,40 @@ THREE.SkinnedMesh.prototype.normalizeSkinWeights = function () {
...
-		// skinning weights assumed to be normalized for THREE.BufferGeometry
+				vec.multiplyScalar( scale );
+
+			} else {
+
+				// do nothing
+
+			}
+
+			skinWeight.setXYZ( i, vec.x, vec.y, vec.z, vec.w )
```
Comment (mrdoob) [inline src/objects/SkinnedMesh.js]: `setXYZW()` maybe?

Comment (WestLangley) [inline src/objects/SkinnedMesh.js]: Oops... Good eyes! : - )

**Comment by titansoftime** [comment]
It seems the ``// do nothing`` are still necessary. Without ``sw.set(1)`` and ``vec.set(1)`` I get some fairly odd visuals:

I will get a fiddle of this up soon.

With ``set(1)``:
set


Without:
donothing

**Comment by WestLangley** [comment]
Please see if you can track down what is going on with your models via the debugger.

`Vector4.set()` takes 4 args, not 1.

**Comment by titansoftime** [comment]
Here's the fiddle: http://jsfiddle.net/cndoe3re/

The ``set(1)`` was just what I took from the original function.

```
THREE.SkinnedMesh.prototype.normalizeSkinWeights = function () {

	if ( this.geometry instanceof THREE.Geometry ) {

		for ( var i = 0; i < this.geometry.skinIndices.length; i ++ ) {

			var sw = this.geometry.skinWeights[ i ];

			var scale = 1.0 / sw.lengthManhattan();

			if ( scale !== Infinity ) {

				sw.multiplyScalar( scale );

			} else {

				sw.set( 1 ); // this will be normalized by the shader anyway

			}

		}

	} else {

		// skinning weights assumed to be normalized for THREE.BufferGeometry

	}

};
```

**Comment by WestLangley** [comment]
Like I said, `Vector4.set( 1 )` is not a proper use of the method. Please see if you can figure out what is going on with your models. I expect you have all-zero weights in places.

**Comment by titansoftime** [comment]
I am not the one who originally put in ``set(1)``, I'm not sure who wrote that, I simply copied it to your ``BufferGeometry`` section of the function.

 I've had this model since r58 and it has been fine until now. So I highly doubt it's the model. I am not a modeler or nor an animator so any investigation on my part would be useless. I can tell you that the weights are not all 0.

http://www.titansoftime.com/utils.php?task=getModel&id=121

http://jsfiddle.net/he21xL1p/ works with the set(1) and  http://jsfiddle.net/cndoe3re/ does not work without it. The function originally had the set(1), it's not of my doing, but it seems to work just fine. It's absence yields breakage. Perhaps the original author of the function could shed some light on this.

**Comment by titansoftime** [comment]
lol @ `` // do something reasonable ``. Works great!

Something was going on other than skin weights of 0 though. You can see here that the weights are not 0:

http://www.titansoftime.com/utils.php?task=getModel&id=121

Regardless, all seems well. Thank you for your efforts =]

**Comment by WestLangley** [comment]
>I expect you have all-zero weights in places.

And you do. Some vertices have all 4 weights equal to zero.

Also, I do not think you give yourself enough credit. I would not have asked for your help if I felt investigation on your part "would be useless" or you were "just a messenger".

**Comment by titansoftime** [comment]
> Some vertices have all 4 weights equal to zero.

Ah, I thought you were saying all weight at 0. Gotcha. A big "duh" on that lol.

> Also, I do not think you give yourself enough credit.

That means a lot, thank you. I feel like you guys are beings from the 5th dimension when it comes to this kind of stuff though =]

**Comment by mrdoob** [comment]
Many thanks guys!

**Comment by MasterJames** [comment]
As a side note every single example anyone has posted since well before I pointed out a month or two ago - continue to show up black ( on Samsung Tab3 chrome current browser ), but I thought we already know that and have addressed it so it must be that people are using older versions or something. Is this expected or did that fix not go through yet?

**Comment by titansoftime** [comment]
The model is black on my tablet due to the fact that I am loading a dds texture which is not supported on my samsung galaxy 4. I suppose I should have just used a png for examples sake.

**Comment by MasterJames** [comment]
Oh I see that explains this one. No it was just the model. The problem before was in the lighting and/or shader and nothing to do with models. I would have expected some visible shading even if the texture didn't load. Is see flat black palm and solid bright green floor on a white BG.

**Pull request merged**